### PR TITLE
Specify vpa resource limit & tune VPA parameters

### PR DESCRIPTION
**Pull request by mrIncompetent**
**What this PR does / why we need it**:
This will:
- Set resource limits on the VPA which ensures that the recommendations are sane (Within valid limits)
- Add a generic function to create VPA's for existing Deployments & StatefulSets
- Tune the VPA parameters to cope a high number of VPA resources

This sets the VPA mode currently to `Auto` so they get used by the e2e tests.
Depending on the test outcome, i'll leave it or deactivate it.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: config/kubermatic/templates/vpa-updater-deployment.yaml [hunk 1]
```diff
@@ -20,6 +20,14 @@ spec:
       containers:
       - name: updater
         image: {{ .Values.kubermatic.vpa.updater.image.repository }}:{{ .Values.kubermatic.vpa.updater.image.tag }}
+        command: ["/updater"]
+        args: [
+          "--address", ":8943",
+          # If a container was killed by a OOM trigger an update.
```
Comment (alvaroaleman) [inline config/kubermatic/templates/vpa-updater-deployment.yaml:27]: What does the timespan mean here?

File: config/kubermatic/templates/vpa-recommender-deployment.yaml [hunk 2]
```diff
@@ -20,6 +20,16 @@ spec:
       containers:
       - name: recommender
         image: {{ .Values.kubermatic.vpa.recommender.image.repository }}:{{ .Values.kubermatic.vpa.recommender.image.tag }}
+        command: ["/recommender"]
+        args: [
+        "--address", ":8942",
+        # Use a higher GC interval to reduce the load on the API.
+        # The recommender only writes new checkpoints when old ones have reached the GC interval
```
Comment (alvaroaleman) [inline config/kubermatic/templates/vpa-recommender-deployment.yaml]: So basically it checks once every hour?

File: config/kubermatic/templates/vpa-admission-controller-deployment.yaml [hunk 3]
```diff
@@ -16,6 +16,11 @@ spec:
       containers:
       - name: admission-controller
         image: {{ .Values.kubermatic.vpa.admissioncontroller.image.repository }}:{{ .Values.kubermatic.vpa.admissioncontroller.image.tag }}
+        command: ["/admission-controller"]
+        args: [
+          "--address", ":8944",
```
Comment (alvaroaleman) [inline config/kubermatic/templates/vpa-admission-controller-deployment.yaml:22]: Why is this needed?

File: api/pkg/resources/vpa.go [hunk 4]
```diff
@@ -17,17 +33,94 @@ func GetVerticalPodAutoscaler(name string, labels map[string]string) VerticalPod
...
+// GetVerticalPodAutoscalersForAll will return functions to create VPA resource for all supplied Deployments and StatefulSets.
+// All resources must exist in the specified namespace.
+// The VPA resource will have the same selector as the Deployment/StatefulSet. The pod container limits will be set as VPA limits.
+func GetVerticalPodAutoscalersForAll(deploymentNames, statefulSetNames []string, namespace string, dynamicCache ctrlruntimecache.Cache) ([]VerticalPodAutoscalerCreator, error) {
+	deploymentStore, err := informer.GetSyncedStoreFromDynamicFactory(dynamicCache, &appsv1.Deployment{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Deployment store: %v", err)
+	}
+
+	creators, err := GetVerticalPodAutoscalersForResource(deploymentNames, namespace, deploymentStore)
```
Comment (alvaroaleman) [inline api/pkg/resources/vpa.go]: please rename to something like `deploymentVPACreators`

File: api/pkg/resources/vpa.go [hunk 5]
```diff
@@ -17,17 +33,94 @@ func GetVerticalPodAutoscaler(name string, labels map[string]string) VerticalPod
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to create VPA creator functions for Deployments: %v", err)
+	}
+
+	statefulSetStore, err := informer.GetSyncedStoreFromDynamicFactory(dynamicCache, &appsv1.StatefulSet{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get StatefulSet store: %v", err)
+	}
+
+	screators, err := GetVerticalPodAutoscalersForResource(statefulSetNames, namespace, statefulSetStore)
```
Comment (alvaroaleman) [inline api/pkg/resources/vpa.go]: Please rename to something like `statefulSetVPACreators`

File: api/pkg/resources/vpa.go [hunk 6]
```diff
@@ -17,17 +33,94 @@ func GetVerticalPodAutoscaler(name string, labels map[string]string) VerticalPod
...
 
 		return pdb, nil
 	}
 }
+
+// GetVerticalPodAutoscalersForAll will return functions to create VPA resource for all supplied Deployments or StatefulSets.
+// If creator functions for VPA's for Deployments should be returned, a deployment store must be passed in. Otherwise a StatefulSet store.
+// All resources must exist in the specified namespace.
+// The VPA resource will have the same selector as the Deployment/StatefulSet. The pod container limits will be set as VPA limits.
+func GetVerticalPodAutoscalersForResource(names []string, namespace string, store cache.Store) ([]VerticalPodAutoscalerCreator, error) {
```
Comment (alvaroaleman) [inline api/pkg/resources/vpa.go]: Why is this public?

File: api/pkg/resources/vpa.go [hunk 7]
```diff
@@ -17,17 +33,94 @@ func GetVerticalPodAutoscaler(name string, labels map[string]string) VerticalPod
 		}
 
 		pdb.Name = name
+		pdb.Namespace = namespace
+		pdb.OwnerReferences = []metav1.OwnerReference{ownerRef}
 
-		updateMode := autoscalingv1beta1.UpdateModeOff
+		updateMode := autoscalingv1beta1.UpdateModeAuto
```
Comment (alvaroaleman) [inline api/pkg/resources/vpa.go:39]: tbh I still don't feel very comfortable with activating this until we have had some more experience with it. Maybe we should add a flag for it, disable by default and get some experience on dev?

File: api/pkg/resources/vpa.go [hunk 8]
```diff
@@ -17,17 +33,94 @@ func GetVerticalPodAutoscaler(name string, labels map[string]string) VerticalPod
...
+// All resources must exist in the specified namespace.
+// The VPA resource will have the same selector as the Deployment/StatefulSet. The pod container limits will be set as VPA limits.
+func GetVerticalPodAutoscalersForResource(names []string, namespace string, store cache.Store) ([]VerticalPodAutoscalerCreator, error) {
+	var creators []VerticalPodAutoscalerCreator
+	for _, name := range names {
+		name := name
+		key := fmt.Sprintf("%s/%s", namespace, name)
+		obj, exists, err := store.GetByKey(key)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get deployment '%s' from store: %v", key, err)
```
Comment (alvaroaleman) [inline api/pkg/resources/vpa.go]: `s/deployment/object`

File: config/kubermatic/templates/vpa-admission-controller-deployment.yaml (hunk 3 above)
Comment (mrIncompetent) [reply config/kubermatic/templates/vpa-admission-controller-deployment.yaml:22]: Not needed, just set explicitly.
Any strong opinion on changing?

File: config/kubermatic/templates/vpa-updater-deployment.yaml (hunk 1 above)
Comment (mrIncompetent) [reply config/kubermatic/templates/vpa-updater-deployment.yaml:27]: That the VPA will scale up (If within limits) if a OOM kill happened within the last 30min

File: config/kubermatic/templates/vpa-recommender-deployment.yaml (hunk 2 above)
Comment (mrIncompetent) [reply config/kubermatic/templates/vpa-recommender-deployment.yaml]: Nope, the GC will cleanup only every hour. 
But that flag does not aid here - i misunderstood the behvaviour of the checkpoint routine.

The recommendee will always update at least 10 VPA checkpoints - and its taking the oldest ones first.
I'll remove that flag

Comment (mrIncompetent) [reply config/kubermatic/templates/vpa-recommender-deployment.yaml]: Removed with https://github.com/kubermatic/kubermatic/pull/2571/commits/effc42c9105aa2dc54199246787417aed420cafc

File: api/pkg/resources/vpa.go (hunk 8 above)
Comment (mrIncompetent) [reply api/pkg/resources/vpa.go]: Done with https://github.com/kubermatic/kubermatic/pull/2571/commits/962d5cb29b2f5647eb15f29a08ff70d06a60d451

File: api/pkg/resources/vpa.go (hunk 5 above)
Comment (mrIncompetent) [reply api/pkg/resources/vpa.go]: Done with https://github.com/kubermatic/kubermatic/pull/2571/commits/962d5cb29b2f5647eb15f29a08ff70d06a60d451

File: api/pkg/resources/vpa.go (hunk 4 above)
Comment (mrIncompetent) [reply api/pkg/resources/vpa.go]: Done with https://github.com/kubermatic/kubermatic/pull/2571/commits/962d5cb29b2f5647eb15f29a08ff70d06a60d451

File: api/pkg/resources/vpa.go (hunk 6 above)
Comment (mrIncompetent) [reply api/pkg/resources/vpa.go]: FIxed with https://github.com/kubermatic/kubermatic/pull/2571/commits/76f645b2ea53f0b9ba66ac88f407ba096eb350b8

File: api/pkg/resources/vpa.go (hunk 7 above)
Comment (mrIncompetent) [reply api/pkg/resources/vpa.go:39]: I introduced a feature flag with https://github.com/kubermatic/kubermatic/pull/2571/commits/75d1e8064d361b24aa3e4c71af2b6ce4f9f406a0

**Comment by mrIncompetent** [comment]
@alvaroaleman PTAL

File: config/kubermatic/templates/vpa-admission-controller-deployment.yaml (hunk 3 above)
Comment (alvaroaleman) [reply config/kubermatic/templates/vpa-admission-controller-deployment.yaml:22]: No its fine, was just wondering

File: config/kubermatic/templates/vpa-updater-deployment.yaml (hunk 1 above)
Comment (alvaroaleman) [reply config/kubermatic/templates/vpa-updater-deployment.yaml:27]: I guess the part that I can't get properly in my mental model of this is: Why is this timeframe needed? I would argue that the VPA should always scale up if an OOM kill happened and that would stay within limits, no matter _when_ the OOM kill happened.

I am just trying to understand the design thinking behind this
